### PR TITLE
Fix #207

### LINKED ISSUE #207: Frame range defined by --frames is not open at the end
Unlike the documentation of --frames, livre_batch.py and previous versions (like https://github.com/BlueBrain/Livre/commit/102cbf257e9a6db50bf27781ff335edffb9efe89), --frames "0 1" generates two images on disk. 

```
EQ_CHANNEL_SATTR_DUMP_IMAGE=/tmp/foo ./livre --frames "0 1" --animation
```
gives foo00000.png and foo00001.png in /tmp.

### PR DESCRIPTION
**Pull request by jafyvilla**
Frame range needs to be open at the end, as the documentation says

### CODE REVIEW

**Comment by jafyvilla** [comment]
Hold it, testing further...

**Comment by jafyvilla** [comment]
Now it should be fine. Tested with memory data source and plugin-based sources (including streaming)

**Comment by rdumusc** [comment]
+1, impossible to judge from the code though :-)

File: livre/eq/Config.cpp [hunk 1]
```diff
@@ -229,11 +229,8 @@ bool Config::frame()
         const uint32_t start = frameMin;
         const uint32_t end = std::max( frameMax, current );
         const int32_t delta = params.animation;
+        const uint32_t interval = end - start;
```
Comment (bilgili) [inline livre/eq/Config.cpp]: are you sure frame number cannot pass the maximum frame id ? Have you tried with the limit ?

Comment (jafyvilla) [inline livre/eq/Config.cpp]: I don't know what you mean with the limit...
Anyways, I'll try to add a test in this commit to finish this madness once and for all.

**Comment by jafyvilla** [comment]
All updated with utility functions (easily 'refactorable' into an animation-specific project when the time comes), and tests. I will add tests in a separate commit to check whether the frame range specified in the command line is correct or not (here I assume it is).

**Comment by bilgili** [comment]
look great, I hope there will never be a corner case again

File: livre/core/util/Utilities.cpp [hunk 2]
```diff
@@ -39,6 +40,41 @@ bool hasOpenGL42Support()
...
+
+    return Vector2ui( frameMin, frameMax );
+}
+
+uint32_t computeCurrentFrameNumber( const uint32_t frameNumber,
+                                    const bool latestAlways,
+                                    const Vector2ui& frameRange )
+{
+    if( latestAlways )
+        return ( frameRange[1] - 1 );
```
Comment (tribal-tec) [inline livre/core/util/Utilities.cpp]: no parentheses needed

**Comment by jafyvilla** [comment]
Yes, me too. But I guess we need to start somewhere...
Don't merge yet though, I want to make sure all our cases are working with this change

File: tests/core/utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    frameRange = { 0, 16 };
+    boundaries = { 8, 23 };
+    BOOST_CHECK_EQUAL( livre::computeFrameRange( frameRange, boundaries ),
+                       livre::Vector2ui( 8, 16 ));
+
+    // outside the boundaries at the end
+    frameRange = { 8, 23 };
+    boundaries = { 0, 16 };
+    BOOST_CHECK_EQUAL( livre::computeFrameRange( frameRange, boundaries ),
+                       livre::Vector2ui( 8, 16 ));
```
Comment (tribal-tec) [inline tests/core/utilities.cpp]: framerange completely outside the boundaries?

Comment (tribal-tec) [inline tests/core/utilities.cpp]: also the trivial test for framerange inside the boundaries.

Comment (jafyvilla) [inline tests/core/utilities.cpp]: Done

File: livre/core/util/Utilities.cpp [hunk 4]
```diff
@@ -39,6 +40,49 @@ bool hasOpenGL42Support()
...
 
+Vector2ui computeFrameRange( const Vector2ui& frameRange,
+                             const Vector2ui& boundaries )
+{
+    // frame range is outside the boundaries
+    if( frameRange[1] <= boundaries[0] || frameRange[0] >= boundaries[1] )
+        return INVALID_FRAME_RANGE;
+
+    if( frameRange[0] < boundaries[0] || frameRange[1] > boundaries[1] )
+        LBVERB << "Clamping the requested frame range to the boundaries"
```
Comment (jafyvilla) [inline livre/core/util/Utilities.cpp]: Added this info, I'm not sure about the level of verbosity... is it correct?

Comment (tribal-tec) [inline livre/core/util/Utilities.cpp]: LBERROR? Not sure. But in any case print the boundaries or the final range that was computed.

File: livre/core/util/Utilities.h [hunk 5]
```diff
@@ -25,6 +25,46 @@ namespace livre
...
+ * delta used as input, as well as the frame range specified.
+ * @param current the current frame number, used as a reference to compute the
+ * next frame number
+ * @param delta the value in which the frame number will be incremented.
+ * It can also be negative
+ * @param frameRange the frame range [x, y), open at the end, in which the
+ * returned value must be contained
+ * @return the next frame number
+ */
+uint32_t computeNextFrameNumber( uint32_t current, int32_t delta,
```
Comment (rdumusc) [inline livre/core/util/Utilities.h]: The documentation for this function should indicate that it loops back to the beginning when the value exceeds the max frame number

**Comment by rdumusc** [comment]
Very good that this frame range is finally unit tested!
I would suggest renaming computeCurrentFrameNumber -> clampFrameNumber and computeNextFrameNumber -> incrementFrameNumber so that it is a bit more generic.

**Comment by bilgili** [comment]
I dont agree on the names. This is not a simple clamping, where the behaviour change later.
It does not always increment the frame number. computeFrame() computeNextFrame() maybe.

**Comment by jafyvilla** [comment]
I updated and refactored the free functions into a FrameUtils class

File: livre/core/types.h [hunk 6]
```diff
@@ -237,6 +237,7 @@ const uint32_t NODEID_FRAME_BITS = 18; //>! @see NodeId
 const uint32_t INVALID_POSITION = ( 1u << NODEID_BLOCK_BITS ) - 1; //!< Invalid node ID.
 const uint32_t INVALID_LEVEL = ( 1u << NODEID_LEVEL_BITS ) - 1; //!< Invalid tree level.4 bits is on
 const uint32_t INVALID_FRAME = ( 1u << NODEID_FRAME_BITS ) - 1; //!< Invalid tree level.4 bits is on
+const uint32_t LATEST_FRAME = INT_MAX;
```
Comment (bilgili) [inline livre/core/types.h]: http://www.cplusplus.com/reference/climits/

If we are getting INT_MAX from here, the limits are as given in the header and it is ushort. If we are getting from somewhere else, be sure that it is INT_MAX. I did accept as si before but the definitions of it is mixed. If we defined it, it is even worse.

File: livre/core/util/FrameUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace livre
+{
+
+FrameUtils::FrameUtils( const Vector2ui& frameRange,
+                        const Vector2ui& boundaries )
+    : _frameRange( INVALID_FRAME_RANGE )
+{
+    // frame range is outside the boundaries
+    if( frameRange[1] <= boundaries[0] || frameRange[0] >= boundaries[1] )
```
Comment (bilgili) [inline livre/core/util/FrameUtils.cpp]: is this correct ? You mean

if( frameRange[ 0 ] < bound[ 0 ] || frameRange[ 0 ] >= bound[ 1 ] ) ?

File: livre/core/util/FrameUtils.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+     * Compute the current frame number, based on the value used as input, as
+     * well as the frame range stored in the object. Optionally, it can make the
+     * returned value stick to the latest available.
+     * @param frameNumber the value used as input to determine the frame number
+     * @param latestAlways if true, return always the last frame number
+     * available: (y - 1)
+     * @return the current frame number. If the active frame range is
+     * INVALID_FRAME_RANGE, return INVALID_FRAME
+     */
+    uint32_t computeCurrentFrameNumber( uint32_t frameNumber,
```
Comment (bilgili) [inline livre/core/util/FrameUtils.h]: since it is FrameUtils getCurrent() getNext() will be fine ?

File: livre/core/util/FrameUtils.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+     * the next frame number
+     * @param delta the value in which the frame number will be incremented.
+     * It can also be negative
+     * @return the next frame number. If the active frame range is
+     * INVALID_FRAME_RANGE, return INVALID_FRAME
+     */
+    uint32_t computeNextFrameNumber( uint32_t current, int32_t delta ) const;
+
+private:
+    Vector2ui _clampFrameRange( const Vector2ui& frameRange,
```
Comment (bilgili) [inline livre/core/util/FrameUtils.h]: _clampToRange() ?

File: livre/eq/Config.cpp [hunk 10]
```diff
@@ -224,24 +220,9 @@ bool Config::frame()
...
-        // If current is at the beginning and animation is reverse,
-        // set current to the end
-        if(( current - start == 0 ) && ( delta < 0 ))
-            current = end;
-
-        const uint32_t frameNumber = (( current - start + delta ) % interval ) + start;
-        frameSettings->setFrameNumber( frameNumber );
-    }
+    if( !keepToLatest )
+        frameSettings->setFrameNumber(
```
Comment (bilgili) [inline livre/eq/Config.cpp]: This place is still a mess  ( not yours, my or others fault ) :) Why do we get the next frame and current frame does not make sense at all. The frame util class could always provide the current frame knowing the previous one. But lets keep it like this now.

File: livre/core/util/FrameUtils.cpp (hunk 7 above)
Comment (jafyvilla) [inline livre/core/util/FrameUtils.cpp]: No, here I mean the case with disjunctive ranges, which is totally invalid.

File: livre/core/types.h (hunk 6 above)
Comment (jafyvilla) [inline livre/core/types.h]: I tried LONG_MAX and ULONG_MAX, but it didn't work, as we are comparing that value with params.animation, which is signed (it can be negative for backwards playback).

File: livre/core/util/FrameUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+Vector2ui FrameUtils::_clampToRange( const Vector2ui& frameRange,
+                                     const Vector2ui& boundaries )
+{
+    const uint32_t frameMin = std::max( frameRange[0], boundaries[0] );
+    const uint32_t frameMax = std::min( frameRange[1], boundaries[1] );
+
+    if( frameRange[0] < boundaries[0] || frameRange[1] > boundaries[1] )
+        LBVERB << "Clamping the requested frame range to the boundaries. "
```
Comment (tribal-tec) [inline livre/core/util/FrameUtils.cpp]: LBVERB is not visible by default. LBINFO is better. It means the user should potentially revisit his/her input of the frame range.

**Comment by tribal-tec** [comment]
Looks very good now! FrameUtils as a class name is weird, but I'm out of any better suggestion right now as well.

File: tests/core/frameUtils.cpp [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    BOOST_CHECK_EQUAL( frameUtils.getCurrent( livre::INVALID_FRAME ), 0 );
+
+    frameRange = { 2, 4 };
+    frameUtils = livre::FrameUtils( frameRange, boundaries );
+    BOOST_CHECK_EQUAL( frameUtils.getCurrent( livre::INVALID_FRAME ), 2 );
+
+    // INVALID_FRAME as output (querying an INVALID_FRAME_RANGE)
+    frameRange = { 0, 8 };
+    boundaries = { 16, 23 };
+    frameUtils = livre::FrameUtils( frameRange, boundaries ); // INVALID_FRAME_RANGE
```
Comment (rdumusc) [inline tests/core/frameUtils.cpp]: Replace comment by:
BOOST_REQUIRE_EQUAL( frameUtils.getFrameRange(), INVALID_FRAME_RANGE )
to validate precondition?

File: livre/eq/Config.cpp (hunk 10 above)
Comment (rdumusc) [inline livre/eq/Config.cpp]: Indeed, that's why I was trying to suggest meaningful names for the functions, like clamp() instead of getFrameCurrent( currentFrame, ... ) :-)

**Comment by jafyvilla** [comment]
ping

**Comment by bilgili** [comment]
pong +1

**Comment by jafyvilla** [comment]
:8ball:

**Comment by jafyvilla** [comment]
Retest this please
